core/chain_indexer.go

```diff
@@ -0,0 +1,296 @@
+	sub := eventMux.Subscribe(ChainEvent{})
+	c.newHead(currentHeader.Number.Uint64(), false)
+	lastHead := currentHeader.Hash()
+	for {
+		select {
+		case <-c.stop:
+			return
+		case ev := <-sub.Chan():
+			header := ev.Data.(ChainEvent).Block.Header()
+			c.newHead(header.Number.Uint64(), header.ParentHash != lastHead)
+			lastHead = header.Hash()
+		}
+	}
+}
+
+// AddChildIndexer adds a child ChainIndexer that can use the output of this one
+func (c *ChainIndexer) AddChildIndexer(ci *ChainIndexer) {
+	c.children = append(c.children, ci)
+}
+
+// newHead notifies the indexer about new chain heads or rollbacks
+func (c *ChainIndexer) newHead(headNum uint64, rollback bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if rollback {
+		firstChanged := headNum / c.sectionSize
+		if firstChanged < c.targetCount {
+			c.targetCount = firstChanged
+		}
+		if firstChanged < c.stored {
+			c.stored = firstChanged
+			c.setValidSections(c.stored)
+		}
+		headNum = firstChanged * c.sectionSize
+
+		if headNum < c.lastForwarded {
+			c.lastForwarded = headNum
+			for _, cp := range c.children {
+				cp.newHead(c.lastForwarded, true)
+			}
+		}
+
+	} else {
+		var newCount uint64
+		if headNum >= c.confirmReq {
+			newCount = (headNum + 1 - c.confirmReq) / c.sectionSize
+			if newCount > c.targetCount {
+				c.targetCount = newCount
+				if !c.updating {
+					c.updating = true
+					c.tryUpdate <- struct{}{}
+				}
+			}
+		}
+	}
+}
+
+// processSection processes an entire section by calling backend functions while ensuring
+// the continuity of the passed headers. Since the chain mutex is not held while processing,
+// the continuity can be broken by a long reorg, in which case the function returns with ok == false.
+func (c *ChainIndexer) processSection(section uint64, lastSectionHead common.Hash) (sectionHead common.Hash, ok bool) {
+	c.backend.Reset(section)
+
+	head := lastSectionHead
+	for i := section * c.sectionSize; i < (section+1)*c.sectionSize; i++ {
+		hash := GetCanonicalHash(c.db, i)
+		if hash == (common.Hash{}) {
+			return common.Hash{}, false
+		}
+		header := GetHeader(c.db, hash, i)
+		if header == nil || header.ParentHash != head {
+			return common.Hash{}, false
+		}
+		c.backend.Process(header)
+		head = header.Hash()
+	}
+	if err := c.backend.Commit(c.db); err != nil {
+		return common.Hash{}, false
+	}
+	return head, true
+}
+
+// CanonicalSections returns the number of processed sections that are consistent with
+// the current canonical chain
+func (c *ChainIndexer) CanonicalSections() uint64 {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	cnt := c.getValidSections()
+	for cnt > 0 {
+		if c.getSectionHead(cnt-1) == GetCanonicalHash(c.db, cnt*c.sectionSize-1) {
+			break
+		}
+		cnt--
+		c.setValidSections(cnt)
+	}
+	return cnt
+}
+
+// getValidSections reads the number of valid sections from the index database
+func (c *ChainIndexer) getValidSections() uint64 {
```

Comment:
It's not two integers. There's one integer at key "count" and one section head hash entry for each section, at key "shead" + sectionIdx (big endian uint64).